File: MainApp/src/main/java/com/sleticalboy/dailywork/weight/xrecycler/decoration/SimpleDecoration.java
package com.sleticalboy.dailywork.weight.xrecycler.decoration;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;
import android.widget.LinearLayout;

/**
 * Created on 18-2-6.
 *
 * @author sleticalboy
 * @version 1.0
 * @description RecyclerView Item 的分割线
 */
public class SimpleDecoration extends RecyclerView.ItemDecoration {

    private static final String TAG = "SimpleDecoration";

    public static final int VERTICAL = LinearLayout.VERTICAL;
    public static final int HORIZONTAL = LinearLayout.HORIZONTAL;

    // 分割线
    private Drawable mDivider;
    // 方向
    private int mOrientation;
    // 边界 left top right bottom
    private final Rect mBounds = new Rect();
    private Context mContext;

    public SimpleDecoration(Context context, int orientation) {
        mContext = context;
        setOrientation(orientation);
    }

    /**
     * Draw any appropriate decorations into the Canvas supplied to the RecyclerView.
     * Any content drawn by this method will be drawn before the item views are drawn,
     * and will thus appear underneath the views.
     *
     * @param canvas Canvas to draw into
     * @param parent RecyclerView this ItemDecoration is drawing into
     * @param state  The current state of RecyclerView
     */
    @Override
    public void onDraw(Canvas canvas, RecyclerView parent, RecyclerView.State state) {
        if (parent.getLayoutManager() == null || mDivider == null) {
            return;
        }
        if (mOrientation == VERTICAL) {
            drawVertical(canvas, parent);
        } else {
            drawHorizontal(canvas, parent);
        }
    }

    private void drawHorizontal(Canvas canvas, RecyclerView parent) {
        canvas.save();
        final int top;
        final int bottom;
        if (parent.getClipToPadding()) { // 判断是否设置了 layout_margin 属性
            top = parent.getPaddingTop();
            bottom = parent.getHeight() - parent.getPaddingBottom();
            canvas.clipRect(parent.getPaddingLeft(), top,
                    parent.getWidth() - parent.getPaddingRight(), bottom);
        } else {
            top = 0;
            bottom = parent.getHeight();
        }
        final int childCount = parent.getChildCount();
        for (int i = 0; i < childCount; i++) {
            final View child = parent.getChildAt(i);
            parent.getDecoratedBoundsWithMargins(child, mBounds);
            final int right = mBounds.right + Math.round(child.getTranslationX());
            final int left = right - mDivider.getIntrinsicWidth();
            mDivider.setBounds(left, top, right, bottom);
            mDivider.draw(canvas);
        }
        canvas.restore();
    }

    private void drawVertical(Canvas canvas, RecyclerView parent) {
        canvas.save();
        final int left;
        final int right;
        if (parent.getClipToPadding()) { // 判断是否设置了 layout_margin 属性
            left = parent.getPaddingLeft();
            right = parent.getWidth() - parent.getPaddingRight();
            canvas.clipRect(left, parent.getPaddingTop(), right,
                    parent.getHeight() - parent.getPaddingBottom());
        } else {
            left = 0;
            right = parent.getWidth();
        }
        final int childCount = parent.getChildCount();
        for (int i = 0; i < childCount; i++) {
            final View child = parent.getChildAt(i);
            parent.getDecoratedBoundsWithMargins(child, mBounds);
            final int bottom = mBounds.bottom + Math.round(child.getTranslationY());
            final int top = mBounds.top - mDivider.getIntrinsicHeight();
            mDivider.setBounds(left, top, right, bottom);
            mDivider.draw(canvas);
        }
        canvas.restore();
    }

    /**
     * Retrieve any offsets for the given item. Each field of <code>outRect</code> specifies
     * the number of pixels that the item view should be inset by, similar to padding or margin.
     * The default implementation sets the bounds of outRect to 0 and returns.
     * <p>
     * <p>
     * If this ItemDecoration does not affect the positioning of item views, it should set
     * all four fields of <code>outRect</code> (left, top, right, bottom) to zero
     * before returning.
     * <p>
     * <p>
     * If you need to access Adapter for additional data, you can call
     * {@link RecyclerView#getChildAdapterPosition(View)} to get the adapter position of the
     * View.
     *
     * @param outRect Rect to receive the output.
     * @param view    The child view to decorate
     * @param parent  RecyclerView this ItemDecoration is decorating
     * @param state   The current state of RecyclerView.
     */
    @Override
    public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State state) {
        if (mDivider == null) {
            outRect.set(0, 0, 0, 0);
            return;
        }
        if (mOrientation == VERTICAL) {
            outRect.set(0, 0, 0, mDivider.getIntrinsicHeight());
        } else {
            outRect.set(0, 0, mDivider.getIntrinsicWidth(), 0);
        }
    }

    public void setOrientation(int orientation) {
        if (orientation != VERTICAL && orientation != HORIZONTAL) {
            throw new IllegalArgumentException(
                    "Invalid orientation. It should be either HORIZONTAL or VERTICAL");
        }
        mOrientation = orientation;
    }

    public void setDrawable(@NonNull Drawable divider) {
        if (divider == null) {
            throw new IllegalArgumentException("Drawable cannot be null.");
        }
        mDivider = divider;
    }

    public void setDrawable(@DrawableRes int drawableId) {
        mDivider = mContext.getResources().getDrawable(drawableId);
    }
}

File: MainApp/src/main/java/com/sleticalboy/weight/xrecycler/decoration/SimpleDecoration.kt
package com.binlee.weight.xrecycler.decoration

import android.content.Context
import android.graphics.Canvas
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.view.View
import android.widget.LinearLayout
import androidx.annotation.DrawableRes
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ItemDecoration

/**
 * Created on 18-2-6.
 *
 * @author leebin
 * @version 1.0
 * @description RecyclerView Item 的分割线
 */
class SimpleDecoration(private val mContext: Context, orientation: Int) : ItemDecoration() {

  // 分割线
  private var mDivider: Drawable? = null

  // 方向
  private var mOrientation = 0

  // 边界 left top right bottom
  private val mBounds = Rect()

  /**
   * Draw any appropriate decorations into the Canvas supplied to the RecyclerView.
   * Any content drawn by this method will be drawn before the item views are drawn,
   * and will thus appear underneath the views.
   *
   * @param canvas Canvas to draw into
   * @param parent RecyclerView this ItemDecoration is drawing into
   * @param state  The current state of RecyclerView
   */
  override fun onDraw(canvas: Canvas, parent: RecyclerView, state: RecyclerView.State) {
    if (parent.layoutManager == null || mDivider == null) {
      return
    }
    if (mOrientation == VERTICAL) {
      drawVertical(canvas, parent)
    } else {
      drawHorizontal(canvas, parent)
    }
  }

  private fun drawHorizontal(canvas: Canvas, parent: RecyclerView) {
    canvas.save()
    val top: Int
    val bottom: Int
    if (parent.clipToPadding) { // 判断是否设置了 layout_margin 属性
      top = parent.paddingTop
      bottom = parent.height - parent.paddingBottom
      canvas.clipRect(
        parent.paddingLeft, top,
        parent.width - parent.paddingRight, bottom
      )
    } else {
      top = 0
      bottom = parent.height
    }
    val childCount = parent.childCount
    for (i in 0 until childCount) {
      val child = parent.getChildAt(i)
      parent.getDecoratedBoundsWithMargins(child, mBounds)
      val right = mBounds.right + Math.round(child.translationX)
      val left = right - mDivider!!.intrinsicWidth
      mDivider!!.setBounds(left, top, right, bottom)
      mDivider!!.draw(canvas)
    }
    canvas.restore()
  }

  private fun drawVertical(canvas: Canvas, parent: RecyclerView) {
    canvas.save()
    val left: Int
    val right: Int
    if (parent.clipToPadding) { // 判断是否设置了 layout_margin 属性
      left = parent.paddingLeft
      right = parent.width - parent.paddingRight
      canvas.clipRect(
        left, parent.paddingTop, right,
        parent.height - parent.paddingBottom
      )
    } else {
      left = 0
      right = parent.width
    }
    val childCount = parent.childCount
    for (i in 0 until childCount) {
      val child = parent.getChildAt(i)
      parent.getDecoratedBoundsWithMargins(child, mBounds)
      val bottom = mBounds.bottom + Math.round(child.translationY)
      val top = mBounds.top - mDivider!!.intrinsicHeight
      mDivider!!.setBounds(left, top, right, bottom)
      mDivider!!.draw(canvas)
    }
    canvas.restore()
  }

  /**
   * Retrieve any offsets for the given item. Each field of `outRect` specifies
   * the number of pixels that the item view should be inset by, similar to padding or margin.
   * The default implementation sets the bounds of outRect to 0 and returns.
   *
   *
   *
   *
   * If this ItemDecoration does not affect the positioning of item views, it should set
   * all four fields of `outRect` (left, top, right, bottom) to zero
   * before returning.
   *
   *
   *
   *
   * If you need to access Adapter for additional data, you can call
   * [RecyclerView.getChildAdapterPosition] to get the adapter position of the
   * View.
   *
   * @param outRect Rect to receive the output.
   * @param view    The child view to decorate
   * @param parent  RecyclerView this ItemDecoration is decorating
   * @param state   The current state of RecyclerView.
   */
  override fun getItemOffsets(
    outRect: Rect,
    view: View,
    parent: RecyclerView,
    state: RecyclerView.State
  ) {
    if (mDivider == null) {
      outRect[0, 0, 0] = 0
      return
    }
    if (mOrientation == VERTICAL) {
      outRect[0, 0, 0] = mDivider!!.intrinsicHeight
    } else {
      outRect[0, 0, mDivider!!.intrinsicWidth] = 0
    }
  }

  private fun setOrientation(orientation: Int) {
    require(!(orientation != VERTICAL && orientation != HORIZONTAL)) { "Invalid orientation. It should be either HORIZONTAL or VERTICAL" }
    mOrientation = orientation
  }

  fun setDrawable(divider: Drawable) {
    mDivider = divider
  }

  fun setDrawable(@DrawableRes drawableId: Int) {
    mDivider = mContext.resources.getDrawable(drawableId)
  }

  companion object {
    private const val TAG = "SimpleDecoration"
    const val VERTICAL = LinearLayout.VERTICAL
    const val HORIZONTAL = LinearLayout.HORIZONTAL
  }

  init {
    setOrientation(orientation)
  }
}